File: array/169.go
func majorityElement(nums []int) int {
    ht := make(map[int]int)
    mark := len(nums) / 2
    for _, v := range nums {
        if count, ok := ht[v]; !ok {
            ht[v] = 1
        }else {
            ht[v]++
            if count + 1 > mark {
                return v
            }
        }
    }
    
    return nums[0]
}

func majorityElement(nums []int) int {
    var (
        num int
        count int
    )
    for _, v := range nums {
        if count == 0 {
            num = v
            count++
        }else {
            if num == v {
                count++
            }else {
                count--
            }
        }
    }
    return num
}

func majorityElement(nums []int) int {
    if len(nums) == 1 {
        return nums[0]
    }
    length := len(nums) / 2
    m := make(map[int]int)
    for _, num := range nums {
        if c, ok := m[num]; !ok {
            m[num] = 1
        }else {
            if c >= length {
                return num
            }
            m[num]++
        }
    }
    return -1
}

File: hashtable/169.cc
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        if (nums.size() == 1) {
            return nums[0];
        }
        std::unordered_map<int, int> check;
        for (const auto& num : nums) {
            auto iter = check.find(num);
            if (iter != check.end()) {
                if ((1+iter->second) > (nums.size()/2)) {
                    return iter->first;
                }
                ++iter->second;
            } else {
                ++check[num];
            }
        }
        return -1;
    }
};
